internal/core/src/common/RangeSearchHelper.cpp

```diff
@@ -0,0 +1,130 @@
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#include <queue>
+#include <vector>
+#include "common/Utils.h"
+
+namespace milvus {
+namespace {
+using MaxHeap = std::priority_queue<std::pair<float, int64_t>,
+                                    std::vector<std::pair<float, int64_t>>,
+                                    std::less<std::pair<float, int64_t>>>;
+using MinHeap = std::priority_queue<std::pair<float, int64_t>,
+                                    std::vector<std::pair<float, int64_t>>,
+                                    std::greater<std::pair<float, int64_t>>>;
+}  // namespace
+
+DatasetPtr
+SortRangeSearchResult(DatasetPtr data_set, int64_t topk, int64_t nq, std::string metric_type) {
+    /**
+     * nq: number of querys;
+     * lims: the size of lims is nq + 1, lims[i+1] - lims[i] refers to the size of RangeSearch result querys[i]
+     *          for example, the nq is 5. In the seleted range,
+     *          the size of RangeSearch result for each nq is [1, 2, 3, 4, 5],
+     *          the lims will be [0, 1, 3, 6, 10, 15];
+     * ids: the size of ids is lim[nq],
+     *      { i(0,0), i(0,1), …, i(0,k0-1),
+     *        i(1,0), i(1,1), …, i(1,k1-1),
+     *          …,
+     *        i(n-1,0), i(n-1,1), …, i(n-1,kn-1)},
+     *      i(0,0), i(0,1), …, i(0,k0-1) means the ids of RangeSearch result querys[0], k0 equals lim[1] - lim[0];
+     * dist: the size of ids is lim[nq],
+     *      { d(0,0), d(0,1), …, d(0,k0-1),
+     *        d(1,0), d(1,1), …, d(1,k1-1),
+     *          …,
+     *        d(n-1,0), d(n-1,1), …, d(n-1,kn-1)},
+     *      d(0,0), d(0,1), …, d(0,k0-1) means the distances of RangeSearch result querys[0], k0 equals lim[1] - lim[0];
+     */
+    auto lims = GetDatasetLims(data_set);
+    auto id = GetDatasetIDs(data_set);
+    auto dist = GetDatasetDistance(data_set);
+
+    // use p_id and p_dist to GenResultDataset after sorted
+    auto p_id = new int64_t[topk * nq];
+    auto p_dist = new float[topk * nq];
+    // cnt means the subscript of p_id and p_dist
+    int cnt = 0;
+
+    for (int i = 0; i < nq; i++) {
+        // if RangeSearch answer size of one nq is less than topk, set the capacity to size
+        int size = lims[i + 1] - lims[i];
+        int capacity = topk > size ? size : topk;
+        /*
+         *   get result for one nq
+         *   IP:   1.0        range_filter     radius
+         *          |------------+---------------|       min_heap   descending_order
+         *   L2:   0.0        range_filter     radius
+         *          |------------+---------------|       max_heap   ascending_order
+         *
+         */
+        if (IsMetricType(metric_type, knowhere::metric::IP)) {
```

Comment:
only the `sub_result` variable should be in if-else, and the for loop should be outside. 
you can try code like this 
```
    std::function<bool(int64_t, int64_t)> cmp;
    if (some condition) {
        cmp = std::less<int64_t>();
    } else {
        cmp = std::greater<int64_t>();
    }
    std::priority_queue<int64_t, std::vector<int64_t>, decltype(cmp)> sub_result;
```
https://stackoverflow.com/questions/36713622/delegate-stdless-or-stdgreater-in-condition